### PR TITLE
[ractor] shareable_constant_value: literal makes constants shareable between ractors

### PR DESCRIPTION
**Pull request by josacar**
Ref. #823
>  Magic comment shareable_constant_value added to freeze constants.
> See {Magic Comments}[rdoc-ref:doc/syntax/comments.rdoc@Magic+Comments] for more details.
> [Feature #17273]

### CODE REVIEW

**Comment by ioquatix** [comment]
I think immutability would be a much better model for users. This is too weak.

File: language/ractor_spec.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
+require_relative '../spec_helper'
+
+describe "Magic comments" do
```
Comment (eregon) [inline language/ractor_spec.rb:3]: These examples seem to clearly come from CRuby tests.
While that in itself is not a problem, it also doesn't improve coverage.
In general, it's much better to start from scratch when writing specs, to discover edge cases and have a structure of tests testing the various possibilities.

File: language/fixtures/shareable_constant_value_magic_comment.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# shareable_constant_value: literal
+
+A = [1, 2]
+H = {a: "a"}
+
+Ractor.new do
+  A
+  H
+en.take
```
Comment (kachick) [inline language/fixtures/shareable_constant_value_magic_comment.rb:9]:
```suggestion
end.take
```

?

**Comment by eregon** [comment]
@josacar Hello, any chance you could finish this PR based on the feedback?

**Comment by josacar** [comment]
I'll try to finish it once I'm back home.

File: language/ractor_spec.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+  ruby_version_is "3.0" do
+    it 'makes constants shareable between ractors' do
+      -> { ruby_exe(fixture(__FILE__, 'shareable_constant_value_magic_comment.rb'), options: '-W0')}.should_not raise_error
```
[the remark this answers is not in the document]
Comment (josacar) [reply language/ractor_spec.rb:6]: Ok, I'll change the constants and check the output. Thanks for helping with the feedback.

File: language/ractor_spec.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        # shareable_constant_value: literal
+        C = [["shareable", "constant#{nil}"]]
+        [A, B, C]
+      end;
+      Ractor.shareable?(a).should == true
+      Ractor.shareable?(b).should == false
+      Ractor.shareable?(c).should == true
+    end
+
+    it 'raises an error whenn shared constant is not a literal' do
```
[the remark this answers is not in the document]
Comment (josacar) [reply language/ractor_spec.rb:25]: Fixed typo, thanks again

File: language/ractor_spec.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+describe "Magic comments" do
+  ruby_version_is "3.0" do
+    it 'makes constants shareable between ractors' do
+      -> { ruby_exe(fixture(__FILE__, 'shareable_constant_value_magic_comment.rb'), options: '-W0')}.should_not raise_error
+    end
+
+    it 'makes constants shareable between ractors' do
+      a, b, c = Class.new.class_eval("#{<<~"begin;"}\n#{<<~'end;'}")
+      begin;
```
[the remark this answers is not in the document]
Comment (josacar) [reply language/ractor_spec.rb:11]: Fixed, thanks again

File: language/ractor_spec.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        C = [["shareable", "constant#{nil}"]]
+        [A, B, C]
+      end;
+      Ractor.shareable?(a).should == true
+      Ractor.shareable?(b).should == false
+      Ractor.shareable?(c).should == true
+    end
+
+    it 'raises an error whenn shared constant is not a literal' do
+      -> { Class.new.class_eval("#{<<~"begin;"}\n#{<<~'end;'}", /unshareable expression/) }.should raise_error
```
[the remark this answers is not in the document]
Comment (josacar) [reply language/ractor_spec.rb:26]: Fixed, thanks again

File: language/ractor_spec.rb (hunk 1 above)
Comment (josacar) [reply language/ractor_spec.rb:3]: I'm a newbie when using Ractors, so I stole the code from CRuby to have something to start and have the foundation for further cases

File: language/fixtures/shareable_constant_value_magic_comment.rb (hunk 2 above)
Comment (josacar) [reply language/fixtures/shareable_constant_value_magic_comment.rb:9]: Fixed, thanks

**Comment by eregon** [comment]
I will close this one, sorry, because Ractor seems (currently at least) very much an experimental CRuby-specific feature.
The reasoning in more details is in https://github.com/ruby/spec/pull/1043#issuecomment-1930313473
